runners/samza/src/main/java/org/apache/beam/runners/samza/renderer/PipelineJsonRenderer.java

```diff
@@ -0,0 +1,186 @@
+        ServiceLoader.load(SamzaIOInfo.SamzaIORegistrar.class).iterator();
+    return beamIORegistrarIterator.hasNext()
+        ? Iterators.getOnlyElement(beamIORegistrarIterator).getSamzaIO()
+        : null;
+  }
+
+  @Override
+  public void enterPipeline(Pipeline p) {}
+
+  @Override
+  public CompositeBehavior enterCompositeTransform(TransformHierarchy.Node node) {
+    String fullName = node.getFullName();
+    writeLine("{ \"fullName\":\"%s\",", assignNodeName(fullName));
+    if (node.getEnclosingNode() != null) {
+      String enclosingNodeName = node.getEnclosingNode().getFullName();
+      writeLine("  \"enclosingNode\":\"%s\",", assignNodeName(enclosingNodeName));
+    }
+
+    Optional<String> ioInfo = getIOTopicInfo(node);
+    if (ioInfo.isPresent() && !ioInfo.get().isEmpty()) {
+      writeLine(" \"ioInfo\":\"%s\",", escapeString(ioInfo.get()));
+    }
+
+    writeLine("  \"ChildNodes\":[");
+    enterBlock();
+    return CompositeBehavior.ENTER_TRANSFORM;
+  }
+
+  @Override
+  public void leaveCompositeTransform(TransformHierarchy.Node node) {
+    exitBlock();
+    writeLine("]},");
+  }
+
+  @Override
+  public void visitPrimitiveTransform(TransformHierarchy.Node node) {
+    String fullName = node.getFullName();
+    writeLine("{ \"fullName\":\"%s\",", escapeString(fullName));
+    String enclosingNodeName = node.getEnclosingNode().getFullName();
+    writeLine("  \"enclosingNode\":\"%s\"},", assignNodeName(enclosingNodeName));
+
+    node.getOutputs().values().forEach(x -> valueToProducerNodeName.put(x, fullName));
+    node.getInputs()
+        .forEach(
+            (key, value) -> {
+              final String producerName = valueToProducerNodeName.get(value);
+              graphLinks.append(
+                  String.format("{\"from\":\"%s\"," + "\"to\":\"%s\"},", producerName, fullName));
+            });
+  }
+
+  @Override
+  public void visitValue(PValue value, TransformHierarchy.Node producer) {}
+
+  @Override
+  public void leavePipeline(Pipeline pipeline) {}
+
+  private void begin() {
+    writeLine("{ \n \"RootNode\": [");
+    graphLinks.append(",\"graphLinks\": [");
+    enterBlock();
+  }
+
+  private void end() {
+    exitBlock();
+    writeLine("]");
+    // delete the last comma
+    int lastIndex = graphLinks.length() - 1;
+    if (graphLinks.charAt(lastIndex) == ',') {
+      graphLinks.deleteCharAt(lastIndex);
+    }
+    graphLinks.append("]");
+    jsonBuilder.append(graphLinks);
+    jsonBuilder.append("}");
+  }
+
+  private void enterBlock() {
+    indent += 4;
+  }
+
+  private void exitBlock() {
+    indent -= 4;
+  }
+
+  private void writeLine(String format, Object... args) {
+    int secondLastCharIndex = jsonBuilder.length() - 2;
+    if (jsonBuilder.length() > 1
+        && jsonBuilder.charAt(secondLastCharIndex) == ','
+        && (format.startsWith("}") || format.startsWith("]"))) {
+      jsonBuilder.deleteCharAt(secondLastCharIndex);
```

Comment:
This condition and the deletion here is very hard to understand. Please add java doc.